### ISSUE TITLE
Anonymous constants should not be included in the Structure View

### ISSUE THREAD

**Issue by afetisov** [intellij-rust/intellij-rust#7221]
## Environment

* **IntelliJ Rust plugin version:** 0.3.146.3826-211
* **Rust toolchain version:** 1.54.0-nightly (bacf770f2 2021-05-05) x86_64-unknown-linux-gnu
* **IDE name and version:** CLion 2021.1 (CL-211.6693.114)
* **Operating system:** Linux 5.4.0-72-generic
* **Macro expansion engine:** new
* **Name resolution engine:** new

## Problem description

Currently anonymous constants (i.e. of the form `const _: T = value;`) are included in the structure view. This is meaningless: such constants cannot be used in any way and exist solely for compile-time assertions. The `static_assertions` crate uses them extensively, and a module with many assertions will be polluted by many meaningless constant definitions.

## Steps to reproduce

Add `static_assertions` crate to the dependencies. In any module write 

```rust
static_assertions::const_assert_eq!(0, 1 - 1);
```

The structure view gets a new entry of the form `null: [{}; {}]`. These nameless entries should not be added.

The reason the nameless constant is added is that the above macro expands to the following code:

```rust
#[allow(unknown_lints, eq_op)]
    const _: [(); 0 - !{
        const ASSERT: bool = (0 == (1 - 1));
        ASSERT
    } as usize] = [];
```
 
which statically checks the equality of terms in the macro.

Similar anonymous constants are added for most other macros in `static_assertions`, such as `assert_eq_size`, `assert_impl_one` or `assert_obj_safe`.

**Comment by Undin** [comment]
According to the [documentation](https://doc.rust-lang.org/reference/items/constant-items.html#unnamed-constant), I can put other items (including `impl`) into const initialization block. For example, the following code works:
```rust
trait Foo {}

struct A;
struct B;

const _: () = {
    impl Foo for A {}
};

const _: () = {
    impl Foo for B {}
};

fn foo<T: Foo>(_x: T) {}

fn main() {
    foo(A);
    foo(B);
}
```
And I suppose, the Structure View should show such items as well as we already do for items inside function blocks.
So, I think we can show only such unnamed constants that contain other items. On one hand, it should clean up the Structure View of meaningless items, on the other hand, show interesting items.
And, of course, fix `null` name in the view

**Comment by afetisov** [comment]
I disagree that such items should be listed in the Structure View. I seriously doubt that such information would be useful. 

Firstly, why would anyone normally put it there in the first place? Structs and traits which are defined within function/const blocks are invisible to the outside world, and can only be accessed via the containing function/const block itself. The common example is to return an impl Trait where the implementing type is local to the function. However, the constant is anonymous, which makes it impossible to use the structs in any way.

Impls are a bit more complicated since one can do a local impl of a public trait for a public type and it will be automatically globally available, but I still don't see any reason to do it within a const block rather than at the global level, and I wouldn't be surprised if impl visibility rules changed in a later edition.

Secondly, let's return to the case of the `static_assertions` crate. It's quite commonly used and I'm sure there are other crates which follow the same pattern.

Consider the macros `assert_fields`, `assert_impl_one`, `assert_impl_all`, `assert_impl_any`, `assert_not_impl_all`, `assert_not_impl_any`, `assert_trait_sub_all`, `assert_type_eq_all` and `assert_type_ne_all`.

All of them define local structs and types, to avoid polluting the global namespace, as well as local impls for those types and structs which perform the actual compile-time checks and type-level computations. However, none of those types and impls are available outside the anonymous const, and none of them are intended to be. Their only purpose is performing compile-time checks.

Considering that `static_assertions` has more than 8,000,000 downloads and more than 250 dependent crates, it would be quite uncomfortable to see all those placeholder types in the Structure View.

At the very least one should require that the types or the traits in the impl are visible outside of the const block for the impl in the Structure View to make any sense (there's also a more complicated case where a local generic type/trait may have public parameters...). But personally I strongly doubt that there are any valid cases where one would want to see items defined in anonymous const blocks. I would suggest making all such items invisible until someone presents a specific use case where it would be undesirable.

**Comment by afetisov** [comment]
In the above comment I'm talking about the code of the form 

```rust
const _ : () = {
    struct Foo;
    trait T {}
    impl T for Foo {}
    fn inner() {}
    
    ()
};
```

I'm certain that none of these items should be included. The code in your comment is less clear-cut.

**Comment by Stzx** [comment]
I support most of  afetisov points.

But for some users they may want to see the structure or features in the constant block. Maybe we should provide a button like `Sort Alphabetically` to provide an optional display.

```rust
const _: bool = false;
```

For purely anonymous constants still filter them.

**Comment by vlad20012** [comment]
@afetisov all your points can be applied to items inside function bodies, but we show them in the `Structure View`. Java's and Kotlin's `Structure View` does the same. I don't think that the fact that some macros use such constants extensively is a good reason for hiding them. We should make more generic and consistent solutions

**Comment by afetisov** [comment]
Solutions should be usable foremost, and generic only if it doesn't inhibit usability. Generic and consistent are mostly a benefit for the implementors, but as a user hundreds of useless entries in the Structure View make it unusable, plain and simple. And a hundred of static assertions in a file are nothing unheard of, especially for autogenerated files.

The situation with functions is also different in several ways. First, nested items in functions are extremely rare. I struggle to name projects where they even exist (although I've certainly seen some), and I have never encountered them as a navigation issue.

Second, they are different because nested items can be used in many ways, They can be returned from functions (hidden behind an impl Trait or dyn Trait, but the actual contents and the trait impl certainly can matter outside the function). They can also be placed in external structures or passed via FFI into some other code.

The items in const blocks can never be used explicitly if consts are anonymous, and they certainly can't be stored in any way outside of the block, since those cannot change the environment.

The points above do not apply to feature-gated "inline consts" (i.e. constexprs), those can return their inner types, only to explicit bindings to a `_`-named constant.
